### ISSUE TITLE
AdamG has serious numerical instabilities

### ISSUE THREAD

**Issue by Vectorrent** [kozistr/pytorch_optimizer#305]
## Describe the bug

AdamG is completely unusable, so far as I can tell. After several hours of experimenting with hyperparameters, and making adjustments to the source code - I have been unsuccessful in finding a resolution to the following problems:

No matter what hyperparameters you use with AdamG, or what model architecture you use - the gradients/losses will explode somewhere around optimizer step 100-200. See the code below; if you run this, losses will explode at step 147.

Replace the optimizer with a different one, and training will be perfectly stable.

## To Reproduce

```py
from typing import Tuple

import numpy as np
import torch
import torch.nn as nn
from pytorch_optimizer import CosineAnnealingWarmupRestarts, create_optimizer


class ComplexModel(nn.Module):
    def __init__(self, input_dim: int = 100, hidden_dim: int = 256):
        super().__init__()

        # Complex architecture with potential for instability
        self.layer1 = nn.Linear(input_dim, hidden_dim)
        self.layer2 = nn.Linear(hidden_dim, hidden_dim)
        self.layer3 = nn.Linear(hidden_dim, hidden_dim)
        self.layer4 = nn.Linear(hidden_dim, hidden_dim)
        self.output = nn.Linear(hidden_dim, 1)

        self.activation = nn.GELU()  # Non-linear activation
        self.dropout = nn.Dropout(0.1)

        # Initialize with slightly larger weights to stress test optimizer
        for m in self.modules():
            if isinstance(m, nn.Linear):
                nn.init.xavier_normal_(m.weight, gain=1.5)
                if m.bias is not None:
                    m.bias.data.fill_(0.1)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # Complex forward pass with skip connections
        h1 = self.activation(self.layer1(x))
        h2 = self.activation(self.layer2(h1)) + h1
        h3 = self.activation(self.layer3(h2)) + h2
        h4 = self.activation(self.layer4(h3)) + h3
        return self.output(h4)


def generate_data(
    batch_size: int = 32, input_dim: int = 100
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Generate synthetic data with complex patterns."""
    # Create input data
    X = torch.randn(batch_size, input_dim)

    # Create target with non-linear relationship
    y = (
        torch.sin(X.sum(dim=1) * 0.1)
        + torch.square(X[:, 0]) * 0.1
        + torch.exp(-torch.abs(X[:, 1])) * 0.5
    )

    # Add some noise
    y += torch.randn_like(y) * 0.1

    # Reshape y to match model output
    y = y.view(-1, 1)

    return X, y


def train_loop(
    model: nn.Module,
    optimizer: torch.optim.Optimizer,
    n_steps: int = 10000,
    batch_size: int = 32,
    input_dim: int = 100,
    log_interval: int = 100,
) -> list:
    """Training loop that tracks losses and can detect instability."""
    model.train()
    criterion = nn.MSELoss()
    losses = []

    try:
        for step in range(n_steps):
            # Generate fresh data each step
            X, y = generate_data(batch_size, input_dim)

            # Forward pass
            optimizer.zero_grad()
            output = model(X)
            loss = criterion(output, y)

            # Backward pass
            loss.backward()

            # Check for exploding gradients
            grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), float("inf"))
            if torch.isnan(grad_norm) or torch.isinf(grad_norm):
                print(f"Gradient explosion detected at step {step}!")
                print(f"Gradient norm: {grad_norm}")
                break

            optimizer.step()

            # Log progress
            if step % log_interval == 0:
                print(
                    f"Step {step}, Loss: {loss.item():.6f}, Grad norm: {grad_norm:.6f}"
                )

            losses.append(loss.item())

            # Early stopping for extreme instability
            if loss.item() > 1e6:
                print(f"Loss explosion detected at step {step}!")
                break

    except RuntimeError as e:
        print(f"Runtime error at step {step}: {str(e)}")

    return losses


# Usage example:
if __name__ == "__main__":
    # Set random seeds for reproducibility
    torch.manual_seed(42)
    np.random.seed(42)

    # Create model
    model = ComplexModel()

    # Create optimizer (to be replaced with AdamG)
    optimizer = create_optimizer(
        model,
        optimizer_name="AdamG",
        lr=1e-3,
        weight_decay=0.001,
        weight_decouple=True,
        p=0.5,
        q=0.24,
        eps=1e-8,
    )

    # Train
    losses = train_loop(model, optimizer)

    print(f"Final loss: {losses[-1]:.6f}")
```

## Log

Output from running this code:
```
Step 0, Loss: 47.922405, Grad norm: 776.451050
Step 100, Loss: 0.527458, Grad norm: 7.351874
Loss explosion detected at step 147!
Final loss: 4159833088.000000
```

## Expected behavior

I would expect AdamG to provide stable convergence, with performance comparable to AdamW.

**Comment by kozistr** [comment]
@Vectorrent, hi. As you mentioned, I can reproduce the divergence with your example. And I additionally tested on the MNIST dataset and found that training went well in the early phase, but the gradient started to explode after some steps. I used the hyperparameters proposed in the paper.

So, I double-checked my implementation to see if there were any bugs or mistakes, but it seems correct as far as I can tell. Please let me know if you notice any errors in my implementation.

![image](https://github.com/user-attachments/assets/f3ceab8e-abca-4265-9944-a88a420ec216)

It seems weird to me either as long as my implementation is correct. So, I think it'd be good to reach out to the authors about this and try to reproduce the results reported in the paper with their settings as closely as possible.

Besides this issue, `AdamG` claims it is more stable than other optimizers like Adam, and AdamW with theoretical and practical examples. However, I don't think this necessarily means it'll outperform or be more stable in general.

Anyway, thanks for the report!

**Comment by Vectorrent** [comment]
I sent an email to the four authors listed in [the original research](https://arxiv.org/abs/2405.04376). Hopefully we'll get a response!

**Comment by kozistr** [comment]
> I sent an email to the four authors listed in [the original research](https://arxiv.org/abs/2405.04376). Hopefully we'll get a response!

cool! looking forward to their reply :)

**Comment by Vectorrent** [comment]
Alright, we have a response, and a fix! Here is the email I got:
```
Hi Ryan, 

Thank you for your interest in our work. I reviewed your original implementation from this repository (https://github.com/kozistr/pytorch_optimizer/blob/main/pytorch_optimizer/optimizer/adamg.py) and was able to reproduce the issues you mentioned. 
Upon investigation, I identified an implementation error in the code. I've attached a modified version that addresses the bugs. 

The primary modification involves updating: 

line-91: bias_correction1: float = 1.0 - self.debias(beta1, group['step']) 
to 
line-91: bias_correction1: float = self.debias(beta1, group['step’]). 

If you refer to the base class (BaseOptimizer), you'll notice that self.debias already includes the computation $1 - beta1 ** group['step']$. Hence, the extra $1 - ?$ outside is unnecessary and causes the issues mainly. 
In addition, I corrected the same issue for bias_correction2 and updated the hyperparameters to $lr = 1.0$ and $p = 0.2$, aligning with the experimental settings in our work. 

Note that issue of bias_correction causes loss explosion and $lr=0.001$ causes slow convergence.

I hope this resolves the issues you encountered. Best of luck with your testing! 

Best regards, 
Yijiang Pang
```
See the linked PR for the changes to code.

**Comment by kozistr** [comment]
@Vectorrent wow. thanks for that. I should have checked that part closely
